sys/mips/82371AB.c

```diff
@@ -0,0 +1,153 @@
+/* 82371AB PCI-ISA bridge driver */
+
+#define KL_LOG KL_DEV
+#include <sys/device.h>
+#include <sys/rman.h>
+#include <sys/bus.h>
+#include <mips/malta.h>
+#include <sys/pci.h>
+#include <sys/interrupt.h>
+#include <dev/piixreg.h>
+#include <dev/isareg.h>
+#include <sys/libkern.h>
+#include <sys/devclass.h>
+
+#define IO_ISASIZE 1024
+
+#define ICU_ADDR(x) ((x) + 0)
+#define ICU_DATA(x) ((x) + 1)
+#define ICU1_ADDR ICU_ADDR(IO_ICU1)
+#define ICU1_DATA ICU_DATA(IO_ICU1)
+#define ICU2_ADDR ICU_ADDR(IO_ICU2)
+#define ICU2_DATA ICU_DATA(IO_ICU2)
+
+#define ISA_IRQ_BASE (0)
+
+typedef struct intel_isa_state {
+  resource_t *io;
+  rman_t io_rman;
+} intel_isa_state_t;
+
+static resource_t *intel_isa_alloc_resource(device_t *dev, res_type_t type,
+                                            int rid, rman_addr_t start,
+                                            rman_addr_t end, size_t size,
+                                            rman_flags_t flags) {
+  assert(dev->bus == DEV_BUS_ISA);
+
+  rman_t *rman = NULL;
+  intel_isa_state_t *isa = dev->parent->state;
+  bus_space_handle_t bh = 0;
+
+  if (type == RT_IOPORTS) {
+    assert(start >= IO_ICUSIZE);
+    rman = &isa->io_rman;
+    bh = isa->io->r_bus_handle;
+  } else {
+    /* We don't know how to allocate this resource, but maybe someone above us
+     * does. In current context this is necessary to allocate interrupts,
+     * which aren't managed by this driver. */
+    if (dev->parent)
+      return bus_alloc_resource(dev->parent, type, rid, start, end, size,
+                                flags);
+    else
+      panic("Can't allocate resource, no dispatch available!");
+  }
+
+  resource_t *r =
+    rman_reserve_resource(rman, type, rid, start, end, size, 0, flags);
+
+  if (r == NULL)
+    return NULL;
+  r->r_rid = rid;
+
+  if (type != RT_IRQ) {
```

Comment:
Please remove the whole if statement an place the bare body of the statement (at this point `type` is always equal to `RT_IOPORTS`).

Reply:
Done.